components/shared/net/request.rs

```diff
@@ -123,11 +124,155 @@ pub enum ResponseTainting {
+}
+
+impl RequestClient {
+    /// <https://html.spec.whatwg.org/multipage/#consume-a-preloaded-resource>
+    pub fn consume_preloaded_resource(
+        &self,
+        request: &Request,
+        on_response_available: impl FnOnce(Response),
+    ) -> bool {
+        // Step 1. Let key be a preload key whose URL is url,
+        // destination is destination, mode is mode, and credentials mode is credentialsMode.
+        let key = PreloadKey {
+            url: request.url().clone(),
+            destination: request.destination,
+            mode: request.mode.clone(),
+            credentials_mode: request.credentials_mode,
+        };
+        // Step 2. Let preloads be window's associated Document's map of preloaded resources.
+        let mut preloaded_resources_lock = self.preloaded_resources.lock();
+        let preloads = preloaded_resources_lock.as_mut().unwrap();
+        // Step 4. Let entry be preloads[key].
+        let Some(entry) = preloads.get_mut(&key) else {
+            // Step 3. If key does not exist in preloads, then return false.
+            return false;
+        };
+        // Step 5. Let consumerIntegrityMetadata be the result of parsing integrityMetadata.
+        let consumer_integrity_metadata =
+            csp::parse_subresource_integrity_metadata(&request.integrity_metadata);
+        // Step 6. Let preloadIntegrityMetadata be the result of parsing entry's integrity metadata.
+        let preload_integrity_metadata =
+            csp::parse_subresource_integrity_metadata(&entry.integrity_metadata);
+        // Step 7. If none of the following conditions apply:
+        if !(
+            // consumerIntegrityMetadata is no metadata;
+            consumer_integrity_metadata == csp::SubresourceIntegrityMetadata::NoMetadata
+            // consumerIntegrityMetadata is equal to preloadIntegrityMetadata; or
+            || consumer_integrity_metadata == preload_integrity_metadata
+        ) {
+            // then return false.
+            return false;
+        }
+        // Step 10. Otherwise, call onResponseAvailable with entry's response.
+        if let Some(response) = entry.response.as_ref() {
+            on_response_available(response.clone());
+        } else {
+            // Step 9. If entry's response is null, then set entry's on response available to onResponseAvailable.
+            // TODO
+        }
+        // Step 8. Remove preloads[key].
+        //
+        // Moved down to avoid double borrow on preloads with entry
+        preloads.shift_remove(&key);
+        // Step 11. Return true.
+        true
+    }
+}
+
+/// <https://html.spec.whatwg.org/multipage/#system-visibility-state>
+#[derive(Clone, Copy, Default, MallocSizeOf, PartialEq)]
+pub enum SystemVisibilityState {
+    #[default]
+    Hidden,
+    Visible,
+}
+
+/// <https://html.spec.whatwg.org/multipage/#traversable-navigable>
+#[derive(Clone, Copy, Default, MallocSizeOf, PartialEq)]
+pub struct TraversableNavigable {
```

Comment:
I suspect we will want this to live somewhere else and only include a few relevant fields in the net-related data structures (particularly after the TODOs), but I'm fine with dealing with that later.